src/main/scala/org/spartanz/parserz/ParsersModule2.scala

```diff
@@ -0,0 +1,326 @@
+        self._1.zipR(self._2, that)
+    }
+
+    implicit final class ToGrammarOps1[SI, SO, E, A, B](self: Grammar[SI, SO, E, List[A]]) {
+
+      def orEmpty: Grammar[SI, SO, E, List[A]] =
+        self.recover(Nil)
+    }
+
+    implicit final class ToGrammarOps2[SI, SO, E, A: ClassTag, B](self: Grammar[SI, SO, E, Array[A]]) {
+
+      def orEmpty: Grammar[SI, SO, E, Array[A]] =
+        self.recover(Array.empty[A])
+    }
+
+    implicit final class ToGrammarOps3[SI, SO, E, A, B](self: Grammar[SI, SO, E, (A, List[B])]) {
+
+      def foldLeft(fold: (A, B) => A, unfold: A =?> (A, B)): Grammar[SI, SO, E, A] =
+        self.map(
+          arg => {
+            arg._2.foldLeft(arg._1)(fold)
+          },
+          arg => {
+            @tailrec
+            def rec(acc: List[B])(a: A): (A, List[B]) =
+              unfold.lift(a) match {
+                case None          => (a, acc)
+                case Some((a1, b)) => rec(b :: acc)(a1)
+              }
+            rec(Nil)(arg)
+          }
+        )
+    }
+  }
+
+
+
+  private class ParserState[+E](val input: Input, var i: Int, var e: List[InOut.Error])
+
+  final def parser[S, E, A](grammar: Grammar[S, S, E, A]): Input => E \/ A =
+    input => step(grammar, new ParserState(input, 0, Nil))
+
+  private def step[S, E, A](grammar: Grammar[S, S, E, A], ps: ParserState[E]): E \/ A =
+    grammar match {
+      case Grammar.GADT.Consume(c, cond, e) =>
+        import InOut.Consumer.Simple._
+        val cas = c.feed(c.create(), ps.input, ps.i)
+        ps.i += extractCount(cas)
+        try {
+          val res = c.finish(extractState(cas))
+          if (cond == null || cond(res)) Right(res) else Left(e)
+        }
+        catch {
+          case InOut.NoInput  => Left(e)
+        }
+
+      case Grammar.GADT.ConsumeMany(c, e) =>
+        val state = c.create()
+        val count = c.feed(state, ps.input, ps.i)
+        ps.i += count
+        try {
+          Right(c.finish(state))
+        }
+        catch {
+          case InOut.NoInput  => Left(e)
+        }
+
+      case Grammar.GADT.ConsumeToken(c, e) =>
+        val state = c.create()
+        val count = c.feed(state, ps.input, ps.i)
+        ps.i += count
+        try {
+          val res = c.finish(state)
+          if (res == null) Left(e) else Right(res)
+        }
+        catch {
+          case InOut.NotFound => Left(e)
+          case InOut.NoInput  => Left(e)
+        }
+
+
+      case Grammar.GADT.Produce(a)     => Right(a)
+      case Grammar.GADT.Tag(value, _)  => step(value, ps)
+      case Grammar.GADT.Delay(delayed) => step(delayed(), ps)
+
+      case Grammar.GADT.Map(value, to, _) =>
+        step(value, ps).flatMap(to)
+
+      case Grammar.GADT.Filter(value, e, expr) =>
+        step(value, ps).flatMap {
+          a =>
+            if (exprFilter(expr)(a)) Right(a)
+            else Left(e)
+        }
+
+      case zip: Grammar.GADT.Zip[S, S, E, ta, tb] =>
```

Comment:
This should be replaced by `ZipAll` so we can eliminate intermediate allocations (nested tuples).

Reply:
For sure! 
I tested it before, and the overhead in this specific perf test is minimal. Currently I'm trying improvements in isolation but eventually I will merge all the work together.